File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/usecase/UseCasesDependOnlyOnRepositoriesAndUseCases.kt
package com.perrystreet.woof.konsist.usecase

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.constructorParameterTypeNames
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class UseCasesDependOnlyOnRepositoriesAndUseCases : BehaviorSpec() {
    init {
        Given("A UseCase") {
            val useCases = KonsistUtils.useCases

            Then("It only injects repositories and other use cases") {
                useCases.assertTrue(message = Message) { useCase ->
                    useCase.constructorParameterTypeNames.all { it.endsWith("Repository") || it.endsWith("UseCase") }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "UseCases may only depend on Repositories and other UseCases.",
            why = "Use cases compose data access into business rules. Data sources, mappers, or view types would leak other layers into the domain.",
            howToFix = "Inject a Repository that wraps the data source, or another UseCase.",
            badExample = "class SendWoofUseCase(private val dataSource: IWoofsDataSource)",
            goodExample = "class SendWoofUseCase(private val woofsRepository: WoofsRepository)",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UseCase/UseCasesDependOnlyOnRepositoriesAndUseCases.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class UseCasesDependOnlyOnRepositoriesAndUseCases: QuickSpec {
    override class func spec() {
        Given("A UseCase") {
            let useCases = WoofHarmonize.useCases

            Then("It only injects repositories and other use cases") {
                useCases.assertTrue(message: message) { useCase in
                    useCase.variables
                        .filter { $0.isStored && !$0.modifiers.contains(.static) }
                        .allSatisfy { variable in
                            let type = variable.typeAnnotation?.name ?? ""
                            return type.hasSuffix("Repository") || type.hasSuffix("UseCase")
                        }
                }
            }
        }
    }

    private static let message = LintRuleMessage(
        rule: "UseCases may only depend on Repositories and other UseCases.",
        why: "Use cases compose data access into business rules. Data sources, mappers, or view types would leak other layers into the domain.",
        howToFix: "Inject a Repository that wraps the data source, or another UseCase.",
        badExample: "final class SendWoofUseCase { private let dataSource: WoofsDataSourceImplementing }",
        goodExample: "final class SendWoofUseCase { private let woofsRepository: WoofsRepository }"
    )
}
